File: coding-task/scala/src/main/scala/ImmutableQueue.scala
class ImmutableQueue[T](val elements: T*) extends AnyRef with Queue[T] {
  override def isEmpty: Boolean = elements.isEmpty

  override def enQueue(t: T): Queue[T] = new ImmutableQueue[T](elements :+ t: _*)

  override def deQueue(): Queue[T] = new ImmutableQueue[T](elements.tail: _*)

  override def head: Option[T] = elements.headOption

  override def equals(obj: Any): Boolean = {
    obj.isInstanceOf[ImmutableQueue[T]] &&
      (this.hashCode() == obj.asInstanceOf[ImmutableQueue[T]].hashCode())
  }

  override def hashCode(): Int = elements.hashCode()
}

object ImmutableQueue {
  def apply[T](elem: T*): ImmutableQueue[T] = new ImmutableQueue[T](elem: _*)
}

File: coding-task/java/src/main/java/ImmutableQueue.java
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

public class ImmutableQueue<T> implements Queue<T> {

    private List<T> elements;

    ImmutableQueue(List<T> elements) {
        this.elements = elements;
    }

    @Override
    public boolean isEmpty() {
        return elements.isEmpty();
    }

    @Override
    public Queue<T> enQueue(T t) {
        elements.add(t);
        return new ImmutableQueue<>(elements);
    }

    @Override
    public Queue<T> deQueue() {
        if (elements.isEmpty()) {
            throw new NoSuchElementException("dequeue on empty queue");
        }
        return new ImmutableQueue<>(elements.subList(1, elements.size()));
    }

    @Override
    public T head() {
        if (elements.isEmpty()) {
            return null;
        }
        return elements.get(0);
    }

    @Override
    public int hashCode() {
        return elements.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof ImmutableQueue && (this.hashCode() == obj.hashCode())) {
            return true;
        }
        return false;
    }
}
